### DISCUSSION TITLE
Does checking for getOutputCount before writing on the port make sense?

### DISCUSSION THREAD

**Discussion by traversaro** [robotology/yarp#2533] [Support - Q&A]
Hi YARP developers and users, 
  I have a doubt on the [`BufferPort::getOutputCount`](https://www.yarp.it/v3.4/classyarp_1_1os_1_1BufferedPort.html#aeda71e41212d06a9f08efa28e3f27233) function. A pattern quite widespread in some YARP-related codebases (see for example icub-main, i.e. https://github.com/robotology/icub-main/search?q=getOutputCount) that I also propagated over the years is to write on a BufferPort only if `getOutputCount()` is greater then zero:
~~~cxx
template <class T> void broadcastData(T& _values, yarp::os::BufferedPort<T>& _port)
{
    if (_port.getOutputCount()>0 )
    {
        _port.prepare()  = _values ;
        _port.write();
    }
}
~~~

I never found this discussed, but I can imagine that the goal is to reduce the use of CPU time (or other resources used in `_port.prepare()/_port.write()` calls) by avoiding those calls if no one is connected to the specific port. The hidden assumption here is that `_port.getOutputCount()` is rather cheap call, i.e. something that is just accessing a small buffer. 

However, if you wear your muddy boots and check the code in PortCore, it seems that `getOutputCount` also calls the `cleanUnits(false)` method:
https://github.com/robotology/yarp/blob/9b52a71e176e23ebe42956cf5b23d8ae40b255f0/src/libYARP_os/src/yarp/os/impl/PortCore.cpp#L1428

That in turns seems a rather complicated function (and to be honest I am not sure what it is supposed to do):
https://github.com/robotology/yarp/blob/9b52a71e176e23ebe42956cf5b23d8ae40b255f0/src/libYARP_os/src/yarp/os/impl/PortCore.cpp#L591

If you check the `cleanUnits` docs, it says "// called anytime, garbage collects terminated units":
> https://github.com/robotology/yarp/blob/9b52a71e176e23ebe42956cf5b23d8ae40b255f0/src/libYARP_os/src/yarp/os/impl/PortCore.h#L606

And while I am not sure what that means in this context, "garbage collection" is not something that is typically associated with inexpensive operations.

Having said that, my questions is:

**TL;DR: Does it make sense use getOutputCount to avoid writing on a BufferedPort, or perhaps it is better to just write all the time on the BufferPort?**

**Comment by traversaro** [comment]
Related issues (that actually sparkled this discussion):
* https://github.com/robotology/walking-controllers/issues/83
* https://github.com/robotology/yarp/issues/2531

This is quite relevant to the icub-main codebase, in which pattern is quite widespread. @pattacini @Nicogene

**Comment by pattacini** [comment]
I think that using `getOutputCount()` can be still meaningful in some contexts.

Ideally, the function should simply return a number, hence no other instructions should be executed.
In turn, this number should be updated by other methods of the port itself when a connection/disconnection event occurs.

**Comment by drdanz** [comment]
To my understanding, if you use it to avoid writing on the port, this is just wrong, because the port will do the same internally.
On the other hand, if you do it to avoid doing some other (potentially heavy) processing or large copies, it might make sense.
